### PR TITLE
add tfs container support

### PR DESCRIPTION
**Pull request by jesterhazy**
*Description of changes:*

add support for sagemaker-tfs inference container

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have updated the [changelog](https://github.com/aws/sagemaker-python-sdk/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [ ] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/tensorflow/estimator.py [hunk 1]
```diff
@@ -305,18 +317,42 @@ def create_model(self, model_server_workers=None, role=None, vpc_config_override
                 Default: use subnets and security groups from this Estimator.
                 * 'Subnets' (list[str]): List of subnet ids.
                 * 'SecurityGroupIds' (list[str]): List of security group ids.
+            endpoint_type: Optional. Selects the software stack used by the inference server.
+                If  not specified, the model will be configured to use the default SageMaker model server.
+                If 'tfs', the model will be configured to use the SageMaker Tensorflow Serving container.
```
[the remark this answers is not in the document]
Comment (jesterhazy) [reply src/sagemaker/tensorflow/estimator.py]: yes, i am planning to add full set of docs here and in the container package in follow-up PR

File: src/sagemaker/predictor.py [hunk 2]
```diff
@@ -57,35 +57,27 @@ def __init__(self, endpoint, sagemaker_session=None, serializer=None, deserializ
...
-        if self.accept:
-            request_args['Accept'] = self.accept
 
+        request_args = self._create_request_args(data, initial_args)
         response = self.sagemaker_session.sagemaker_runtime_client.invoke_endpoint(**request_args)
+        return self._handle_response(response)
 
+    def _handle_response(self, response):
         response_body = response['Body']
         if self.deserializer is not None:
```
Comment (mvsusp) [inline src/sagemaker/predictor.py]:
```suggestion
        if self.deserializer:
```

File: src/sagemaker/predictor.py [hunk 3]
```diff
@@ -94,6 +86,24 @@ def predict(self, data):
         response_body.close()
         return data
 
+    def _create_request_args(self, data, initial_args=None):
+        args = dict(initial_args) if initial_args else {}
```
Comment (mvsusp) [inline src/sagemaker/predictor.py]:
```suggestion
        args = dict(initial_args or {})
```

File: src/sagemaker/predictor.py [hunk 4]
```diff
@@ -94,6 +86,24 @@ def predict(self, data):
...
+            args['ContentType'] = self.content_type
+
+        if self.accept and 'Accept' not in args:
+            args['Accept'] = self.accept
+
+        if self.serializer is not None:
+            data = self.serializer(data)
+
+        args['Body'] = data
+        return args
```
Comment (mvsusp) [inline src/sagemaker/predictor.py]: nit: You can simplify the code here:

args['ContentType'] = args.get('ContentType', self.content_type)
args['Accept'] = args.get('Accept', self.accept)
args['Body'] = self. serializer(data) # we can provide a default serializer that just return the data itself.

File: src/sagemaker/tensorflow/__init__.py [hunk 5]
```diff
@@ -29,3 +29,4 @@
 
 from sagemaker.tensorflow.estimator import TensorFlow  # noqa: E402, F401
 from sagemaker.tensorflow.model import TensorFlowModel, TensorFlowPredictor  # noqa: E402, F401
+from sagemaker.tensorflow.tfs import TFSModel, TFSPredictor  # noqa: E402, F401
```
Comment (mvsusp) [inline src/sagemaker/tensorflow/__init__.py]: Let's avoid acronyms to keep the same pattern of the SDK

File: CHANGELOG.rst [hunk 6]
```diff
@@ -2,6 +2,11 @@
 CHANGELOG
 =========
 
+1.14.0-dev
+==========
+
+* add support for sagemaker-tfs container
```
Comment (mvsusp) [inline CHANGELOG.rst]: Please follow correct changelog format.

File: src/sagemaker/tensorflow/estimator.py [hunk 7]
```diff
@@ -305,18 +317,42 @@ def create_model(self, model_server_workers=None, role=None, vpc_config_override
...
+                                          vpc_config_override=vpc_config_override)
+
+    def _create_tfs_model(self, role=None, vpc_config_override=VPC_CONFIG_DEFAULT):
+        return TFSModel(model_data=self.model_data,
+                        role=role,
+                        image=self.image_name,
+                        name=self._current_job_name,
+                        framework_version=self.framework_version,
+                        sagemaker_session=self.sagemaker_session,
+                        vpc_config=self.get_vpc_config(vpc_config_override))
```
Comment (mvsusp) [inline src/sagemaker/tensorflow/estimator.py]: Are you missing container_log_level?

File: src/sagemaker/tensorflow/__init__.py (hunk 5 above)
Comment (jesterhazy) [reply src/sagemaker/tensorflow/__init__.py]: 'TensorFlowServingModel' is unwieldy, and ambiguous since TensorFlowModel also refers to a kind of TensorFlow Serving model. TFS is well aligned with the module name, and the container side package name.

Another option... just calling them Model and Predictor (sagemaker.tensorflow.tfs.Model etc), but this invites collision with sagemaker.Model etc.

File: src/sagemaker/tensorflow/estimator.py (hunk 7 above)
Comment (jesterhazy) [reply src/sagemaker/tensorflow/estimator.py]: yes

File: CHANGELOG.rst (hunk 6 above)
Comment (jesterhazy) [reply CHANGELOG.rst]: ok

File: src/sagemaker/tensorflow/tfs.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        return env
+
+    def _get_image_uri(self, instance_type):
+        if self.image:
+            return self.image
+
+        # reuse standard image uri function, then strip unwanted python component
+        region_name = self.sagemaker_session.boto_region_name
+        image = create_image_uri(region_name, TFSModel.FRAMEWORK_NAME, instance_type,
+                                 self._framework_version, 'py3')
```
Comment (mvsusp) [inline src/sagemaker/tensorflow/tfs.py]: How hard is to change create image uri instead of reusing and striping it.

File: tests/integ/test_tfs.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+@pytest.fixture(scope='module')
+def tfs_predictor(sagemaker_session, tf_full_version):
+    endpoint_name = sagemaker.utils.name_from_base('sagemaker-tfs')
+    model_data = sagemaker_session.upload_data(path='tests/data/tfs-test-model.tar.gz',
+                                               key_prefix='tfs/models')
+    with timeout_and_delete_endpoint_by_name(endpoint_name, sagemaker_session):
+        model = TFSModel(model_data, 'SageMakerRole', framework_version=tf_full_version)
+        predictor = model.deploy(1, 'ml.c5.xlarge', endpoint_name=endpoint_name)
+        yield predictor
```
Comment (mvsusp) [inline tests/integ/test_tfs.py]: Nice test setup

File: src/sagemaker/predictor.py (hunk 4 above)
Comment (jesterhazy) [reply src/sagemaker/predictor.py]: nicer, but behavior is diff't because it misses the check for self.content_type etc.
this is mostly pre-existing code (just moved) and i don't want to alter the behavior

File: tests/integ/test_tfs.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+def test_predict_csv(tfs_predictor):
+    input_data = '1.0,2.0,5.0\n1.0,2.0,5.0'
+    expected_result = {'predictions': [[3.5, 4.0, 5.5], [3.5, 4.0, 5.5]]}
+
+    predictor = TFSPredictor(tfs_predictor.endpoint, tfs_predictor.sagemaker_session,
+                             serializer=sagemaker.predictor.csv_serializer)
+
+    result = predictor.predict(input_data)
+    assert expected_result == result
```
Comment (mvsusp) [inline tests/integ/test_tfs.py]: We are missing test cases for errors and empty requests.

File: src/sagemaker/predictor.py (hunk 4 above)
Comment (mvsusp) [reply src/sagemaker/predictor.py]: It's ok if you don't want to change but
```args['ContentType'] = args.get('ContentType', self.content_type)```
and
```
if self.content_type and 'ContentType' not in args:
    args['ContentType'] = self.content_type
```
have the same behavior.

File: src/sagemaker/tensorflow/__init__.py (hunk 5 above)
Comment (mvsusp) [reply src/sagemaker/tensorflow/__init__.py]: what about sagemaker.tesorflow.serving.Model or sagemaker.tensorflow_serving.Model?

Collision should not happen if we are importing modules not classes like specified in the style guide.

File: src/sagemaker/predictor.py (hunk 4 above)
Comment (owen-t) [reply src/sagemaker/predictor.py]: No, not quite.

In the former, you're going to always assign args['ContentType'] a value, which can be None.

Comment (mvsusp) [reply src/sagemaker/predictor.py]: That is correct, my suggestion changes behavior. Thanks for clarifying.

File: tests/integ/test_tfs.py (hunk 10 above)
Comment (jesterhazy) [reply tests/integ/test_tfs.py]: - handled in unit tests and in container package tests
- also this is generic predictor / serializer behavior that is well covered by other tests

File: src/sagemaker/tensorflow/tfs.py (hunk 8 above)
Comment (jesterhazy) [reply src/sagemaker/tensorflow/tfs.py]: lots of fussy behavior in create_image_uri. it would be easy to fall out of sync
